File: src/main/java/de/tectoast/emolga/buttons/ButtonListener.java
package de.tectoast.emolga.buttons;

import com.google.common.reflect.ClassPath;
import de.tectoast.emolga.commands.Command;
import net.dv8tion.jda.api.events.interaction.ButtonClickEvent;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Modifier;
import java.util.HashMap;

public abstract class ButtonListener {

    public static final HashMap<String, ButtonListener> listener = new HashMap<>();
    final String name;
    private static final ButtonListener NULL = new ButtonListener("NULL") {
        @Override
        public void process(ButtonClickEvent e, String name) {
            Command.sendToMe("WRONG BUTTON KEY " + e.getComponentId());
        }
    };

    public ButtonListener(String name) {
        this.name = name;
        listener.put(name, this);
    }

    public static void check(ButtonClickEvent e) {
        System.out.println("e.getComponentId() = " + e.getComponentId());
        String[] split = e.getComponentId().split(";");
        listener.getOrDefault(split[0], NULL).process(e, split[1]);
    }

    @SuppressWarnings("UnstableApiUsage")
    public static void init() {
        ClassLoader loader = Thread.currentThread().getContextClassLoader();
        try {
            for (ClassPath.ClassInfo classInfo : ClassPath.from(loader).getTopLevelClassesRecursive("de.tectoast.emolga.buttons")) {
                Class<?> cl = classInfo.load();
                if (cl.getSuperclass().getSimpleName().endsWith("ButtonListener") && !Modifier.isAbstract(cl.getModifiers())) {
                    //System.out.println(classInfo.getName());
                    cl.getConstructors()[0].newInstance();
                }
            }
        } catch (IOException | IllegalAccessException | InstantiationException | InvocationTargetException e) {
            e.printStackTrace();
        }
    }

    public abstract void process(ButtonClickEvent e, String name);
}

File: src/main/kotlin/de/tectoast/emolga/buttons/ButtonListener.kt
package de.tectoast.emolga.buttons

import com.google.common.reflect.ClassPath
import de.tectoast.emolga.commands.Command
import de.tectoast.emolga.commands.isNotFlo
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent
import org.slf4j.LoggerFactory
import java.lang.reflect.Modifier

abstract class ButtonListener(name: String) {
    init {
        if (name !in disabledListeners)
        listener[name] = this
    }

    @Throws(Exception::class)
    abstract suspend fun process(e: ButtonInteractionEvent, name: String)

    companion object {
        val listener: MutableMap<String, ButtonListener> = HashMap()
        val disabledListeners = setOf("reopen")
        private val logger = LoggerFactory.getLogger(ButtonListener::class.java)
        private val NULL: ButtonListener = object : ButtonListener("NULL") {
            override suspend fun process(e: ButtonInteractionEvent, name: String) {
                if (!e.componentId.first().isDigit() && disabledListeners.none { e.componentId.startsWith(it) })
                    Command.sendToMe("WRONG BUTTON KEY " + e.componentId)
            }
        }

        suspend fun check(e: ButtonInteractionEvent) {
            logger.info("e.getComponentId() = {}", e.componentId)
            if (Command.BOT_DISABLED && e.user.isNotFlo) e.reply(Command.DISABLED_TEXT).setEphemeral(true).queue()
            e.componentId.split(";").let {
                listener.getOrDefault(it[0], NULL).process(e, it.drop(1).joinToString(";"))
            }
        }

        fun init() {
            val loader = Thread.currentThread().contextClassLoader
            for (classInfo in ClassPath.from(loader).getTopLevelClassesRecursive("de.tectoast.emolga.buttons")) {
                val cl = classInfo.load()
                if (cl.superclass.simpleName.endsWith("ButtonListener") && !Modifier.isAbstract(cl.modifiers)) {
                    cl.kotlin.objectInstance
                }
            }
        }
    }
}
